### PR TITLE
Some Tweaks to the extended authentication token handling

### PR DESCRIPTION
**Pull request by bwaidelich**
* Rename `UsernamePasswordInterface` to `UsernamePasswordTokenInterface` and make
  it extend the existing `TokenInterface`
* Adjust affected Providers so that they only specify the interface of the
  tokens they support
* Adjust `TokenAndProviderFactory` such that it can deal with the interface
  creation
* Introduce `PasswordTokenInterface` in the same manner

### CODE REVIEW

**Comment by bwaidelich** [comment]
FYI: Successfully tested this with http basic auth, regular username/password and file based (password)  authentication

File: Neos.Flow/Classes/Security/Authentication/TokenAndProviderFactory.php [hunk 1]
```diff
@@ -160,14 +167,13 @@ protected function buildProvidersAndTokensFromConfiguration()
...
-
-                    if ($tokenObjectName !== $tokenClassName) {
-                        continue;
-                    }
+                    $tokenClassName = $this->tokenResolver->resolveTokenClass((string)$providerConfiguration['token']);
+                }
+                /** @noinspection PhpMethodParametersCountMismatchInspection */
+                $tokenInstance = $this->objectManager->get($tokenClassName, $providerConfiguration['tokenOptions'] ?? []);
+                if (!$tokenInstance instanceof TokenInterface) {
+                    throw new Exception\InvalidAuthenticationProviderException('TODO', 1585921152);
```
Comment (albe) [inline Neos.Flow/Classes/Security/Authentication/TokenAndProviderFactory.php:175]: TODO

**Comment by albe** [comment]
Okay, I do like the change to interfaces and have the TokenAndProviderFactory get the actual class from object management. But this still somehow goes beyond what @sorenmalling wanted to achive initially I think (correct me if I'm wrong).

Let's try with a User Story:
> As a developer, I want to easily extend UsernamePasswordToken to get the credentials from another place inside the Request, for example the headers, without having to understand and duplicate all the other logic inside the token.

**Comment by bwaidelich** [comment]
> But this still somehow goes beyond what @sorenmalling wanted to achive initially

Right, and I almost feel bad about it.. But I think if we go interfaces we have to go "all in".

> Let's try with a User Story:

Great idea! :)

So with the original PR (#1945) the user story would be satisfied (as long as you are OK with POST requests only).

*But:* You could not use this token unless you create a custom Provider for it, because the provider currently specifies the implementation class names. And you can't even use a sub class using the `token` configuration because of: https://github.com/neos/flow-development-collection/blob/master/Neos.Flow/Classes/Security/Authentication/TokenAndProviderFactory.php#L165

So all in all I still think it would be a great deal if providers could be independant of the token implementation, but I see the issue of this getting larger and larger

**Comment by albe** [comment]
The user story would be fulfilled if the two methods would receive the request as argument and could decide themself how to retrieve the credentials from there. The default UsernamePassword implementation could forward to another private method that extracts both values into `$this->credentials` or sth.

**Comment by albe** [comment]
So, how do we proceed @bwaidelich @sorenmalling ?

**Comment by bwaidelich** [comment]
@albe I'd be fine with the original PR (neos#1945) to be re-opened and merged, my point was only: It won't allow to adjust the behavior without writing a custom Auth provider, too. That's why I'm not sure if it's worth it.

**Comment by albe** [comment]
Indeed, that's why we also need your `TokenInterface` change with that :) Then you can just configure your own `MyUsernamePasswordToken extends UsernamePasswordToken` via Objects.yaml and be done. It brings flexibility at no additional cost on top of your version here. If we achieve that by adding the `extractCredentialsFromRequest` method from the original PR or by adding the request as argument to the two `get*` methods is not a concern for me.

**Pull request merged**